rust/src/base.rs

```diff
@@ -0,0 +1,205 @@
+                    "cannot load base without a lon value",
+                ))
+            }
+        };
+        let lon = lon_string.parse().map_err(|e| PyValueError::new_err(e))?;
+
+        let stall_count_string = match row.get(&"stall_count".to_string()) {
+            Some(sc) => sc,
+            None => {
+                return Err(PyValueError::new_err(
+                    "cannot load base without a stall_count value",
+                ))
+            }
+        };
+        let stall_count = stall_count_string
+            .parse::<usize>()
+            .map_err(|e| PyValueError::new_err(e))?;
+
+        let h3cell =
+            H3Cell::from_coordinate(coord! {x: lon, y: lat}, road_network.sim_h3_resolution)
+                .map_err(|e| PyValueError::new_err(e.to_string()))?;
+
+        let station_id = match row.get("station_id") {
+            Some(sid) => sid,
+            None => "none",
+        }
+        .to_string();
+
+        Ok(Base {
+            id: Arc::new(base_id),
+            position: Arc::new(road_network.position_from_geoid(h3cell.to_string())),
+            membership: Arc::new(Membership::default()),
+            total_stalls: Arc::new(stall_count),
+            available_stalls: Arc::new(stall_count),
+            station_id: Arc::new(Some(station_id)),
+        })
+    }
+
+    pub fn has_available_stall(&self, membership: Membership) -> bool {
+        *self.available_stalls > 0 && self.membership.grant_access_to_membership(&membership)
+    }
+
+    pub fn checkout_stall(&self) -> Option<Base> {
+        if *self.available_stalls == 0 {
+            return None;
+        } else {
+            let mut new_self = self.clone();
+            let new_stalls = Arc::make_mut(&mut new_self.available_stalls);
+            *new_stalls = *new_stalls - 1;
+            Some(new_self)
+        }
+    }
+
+    pub fn return_stall(&self) -> (Option<PyErr>, Option<Base>) {
+        if (*self.available_stalls + 1) > *self.total_stalls {
+            let err = PyValueError::new_err("base already has max stalls");
+            (Some(err), None)
+        } else {
+            let mut new_self = self.clone();
+            let new_stalls = Arc::make_mut(&mut new_self.available_stalls);
+            *new_stalls = *new_stalls + 1;
+            (None, Some(new_self))
+        }
+    }
+
+    pub fn set_membership(&self, member_ids: Vec<String>) -> PyResult<Base> {
+        let mut new_self = self.clone();
+        let new_membership = Arc::make_mut(&mut new_self.membership);
+        *new_membership = Membership::_from_tuple(member_ids).map_err(PyValueError::new_err)?;
+        Ok(new_self)
+    }
+
+    pub fn add_membership(&self, membership_id: String) -> PyResult<Base> {
+        let mut new_self = self.clone();
+        let new_membership = Arc::make_mut(&mut new_self.membership);
+        *new_membership = new_membership.add_membership(membership_id)?;
+        Ok(new_self)
+    }
+}
```

Comment:
tried at the bottom of this file to build at least a mock_base inside a test module, and i got really stumped. i tried using Python::with_gil( ... ) to call Base::build(...) but couldn't wire it up properly. 🤔

Reply:
In my experience with writing tests with the pyo3 library, I haven't had much success with testing methods that return any python specific types (like a `PyErr` or such). Ideally we would have pure rust objects that we could write unit tests for and then just layer on a thin python api that gets tested from the python side; I tried to demonstrate that with writing a test for the `checkout_stall` method which doesn't return any explicit python types.

Reply:
ya, i had a similar inclination about testing without the Py* types in scope. but then i doubled down, that we should be able to do e2e tests. but maybe those should happen via python test suites that invoke the USE_RUST env variable.